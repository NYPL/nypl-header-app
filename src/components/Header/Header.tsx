import {
  chakra,
  Box,
  HStack,
  useColorModeValue,
  useMultiStyleConfig,
  Spacer,
  VStack,
} from "@chakra-ui/react";

/** Internal Header-only components */
import HeaderLowerNav from "./components/HeaderLowerNav";
import HeaderMobileIconNav from "./components/HeaderMobileIconNav";
import HeaderSitewideAlerts from "./components/HeaderSitewideAlerts";
import HeaderUpperNav from "./components/HeaderUpperNav";
/** Internal Header-only utils */
import { HeaderProvider } from "./context/headerContext";
import { useNYPLBreakpoints, SkipNavigation, Link, Logo, HorizontalRule } from "@nypl/design-system-react-components";

export interface HeaderProps {
  /** Whether to render sitewide alerts or not. True by default. */
  fetchSitewideAlerts?: boolean;
  /** Whether or not the `Header` is in production mode. True by default. */
  isProduction?: boolean;
}

/**
 * The NYPL `Header` component is the top-level component of the site. It
 * contains features for logging in, logging out, searching, and navigating
 * the NYPL.org site.
 */
export const Header = chakra(
  ({
    fetchSitewideAlerts = true,
    isProduction = true,
  }: HeaderProps) => {
    const { isLargerThanMobile, isLargerThanLarge } = useNYPLBreakpoints();
    const styles = useMultiStyleConfig("Header", {});

    return (
      <HeaderProvider isProduction={isProduction}>
        <Box __css={styles}>
          <SkipNavigation />
          {fetchSitewideAlerts ? <HeaderSitewideAlerts /> : null}
          <header>
            <HStack __css={styles.container}>
              <Link
                aria-label="The New York Public Library"
                href="https://nypl.org"
                __css={styles.logo}
              >
                <Logo
                  aria-label="NYPL Header Logo"
                  name={
                    isLargerThanLarge
                      ? useColorModeValue("nyplFullBlack", "nyplFullWhite")
                      : useColorModeValue("nyplLionBlack", "nyplLionWhite")
                  }
                  size={isLargerThanMobile ? "large" : "small"}
                  title="NYPL Header Logo"
                />
              </Link>
              <Spacer />
              {isLargerThanMobile ? (
                <VStack
                  alignItems="end"
                  spacing={isLargerThanLarge ? "85px" : "40px"}
                >
                  <HeaderUpperNav />
                  <HeaderLowerNav />
                </VStack>
              ) : (
                <HeaderMobileIconNav />
              )}
            </HStack>
            <HorizontalRule __css={styles.horizontalRule} />
          </header>
        </Box>
      </HeaderProvider>
    );
  }
);

export default Header;
