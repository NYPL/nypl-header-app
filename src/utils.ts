// This is used just for the app's environment value, either
// qa or production. This function is needed to get around jest
// throwing an error trying to load a cjs module (and this syntax
// is also Vite-specific).
export const getEnvVar = (key: string) => {
  return import.meta.env[key];
};
