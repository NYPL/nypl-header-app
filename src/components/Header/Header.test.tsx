import { render, screen, waitFor, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { axe } from "jest-axe";
import Cookies from "js-cookie";
import * as renderer from "react-test-renderer";

import Header from "./Header";
import {
  patronApiUrlWithToken,
  mockExpiredResponseData,
  mockLoginCookie,
  mockResponseData,
} from "./utils/authApiMockResponse";
import { alertsApiUrl, tokenRefreshLink } from "./utils/headerUtils";
import { refineryResponse } from "./utils/sitewideAlertsMocks";

jest.mock("js-cookie", () => ({
  get: () => mockLoginCookie,
  remove: jest.fn(),
}));

describe("Header Accessibility", () => {
  it("passes axe accessibility test", async () => {
    // Mock the fetch API call in `SitewideAlerts`.
    (global as any).fetch = jest.fn(() =>
      Promise.resolve({
        status: 200,
        json: () => Promise.resolve(refineryResponse),
      })
    ) as jest.Mock;

    const { container } = await waitFor(() =>
      render(<Header isProduction={false} />)
    );
    expect(await axe(container)).toHaveNoViolations();

    jest.clearAllMocks();
  });
});

// TODO: These tests do not currently test the mobile web view.
// We need to determine a way of doing this for all responsive
// components, and will add this in at a later date.
describe.skip("Header", () => {
  let container;

  beforeEach(async () => {
    // Mock the fetch API call in `SitewideAlerts`.
    (global as any).fetch = jest.fn(() =>
      Promise.resolve({
        status: 200,
        json: () => Promise.resolve(refineryResponse),
      })
    ) as jest.Mock;

    // Mocks matchMedia so that the desktop view renders rather than mobile.
    global.matchMedia = jest.fn(() => ({
      addListener: () => {},
      removeListener: () => {},
      matches: true,
    })) as jest.Mock;

    await waitFor(() => {
      const utils = render(<Header isProduction={false} />);
      container = utils.container;
    });
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it("renders a skip navigation", () => {
    const skipNavigation = screen.getAllByRole("navigation")[0];
    const { getByRole } = within(skipNavigation);

    expect(skipNavigation).toBeInTheDocument();
    expect(skipNavigation).toHaveAttribute("aria-label", "Skip Navigation");
    expect(getByRole("list")).toBeInTheDocument();
  });

  it("renders a notification", () => {
    const notification = screen.getByRole("complementary");
    const { getByText } = within(notification);

    expect(notification).toBeInTheDocument();
    expect(getByText(/masks are encouraged/i)).toBeInTheDocument();
  });

  it("renders the NYPL logo", () => {
    const nyplLink = screen.getAllByRole("link", {
      name: "The New York Public Library",
    })[0];

    const logo = within(nyplLink).getByRole("img");

    expect(logo).toHaveAttribute("aria-label", "NYPL Header Logo");
  });

  it("renders the upper links", () => {
    // Removes automatically added, unused Chakra toast elements.
    document.getElementById("chakra-toast-portal")?.remove();

    // The first list is the skip navigation.
    // The second list is the list of alerts in the `SitewideAlerts` component.
    // The third list is the upper navigation.

    const upperList = screen.getAllByRole("list")[2];
    const upperLinks = within(upperList).getAllByRole("listitem");

    expect(upperLinks.length).toEqual(6);
    expect(upperLinks[0]).toHaveTextContent(/log in/i);
    expect(upperLinks[5]).toHaveTextContent(/shop/i);
  });

  it("renders the lower links", () => {
    // Removes automatically added, unused Chakra toast elements.
    document.getElementById("chakra-toast-portal")?.remove();

    // The first list is the skip navigation.
    // The second list is the list of alerts in the `SitewideAlerts` component.
    // The third list is the upper navigation.
    // The fourth list is the lower navigation.
    const lowerList = screen.getAllByRole("list")[3];
    const lowerLinks = within(lowerList).getAllByRole("listitem");

    expect(lowerLinks.length).toEqual(8);
    expect(lowerLinks[0]).toHaveTextContent(/books/i);
    expect(lowerLinks[7]).toHaveTextContent(/search/i);
  });

  it("opens the login menu", () => {
    // Removes automatically added, unused Chakra toast elements.
    document.getElementById("chakra-toast-portal")?.remove();

    // The third list is the upper navigation links.
    let upperList = screen.getAllByRole("list")[2];
    let upperLinks = within(upperList).getAllByRole("listitem");

    expect(upperLinks.length).toEqual(6);

    const myAccountButton = within(upperLinks[0]).getByRole("button");
    expect(upperLinks[0]).toHaveTextContent(/my account/i);
    expect(upperLinks[1]).toHaveTextContent(/locations/i);

    userEvent.click(myAccountButton);

    upperList = screen.getAllByRole("list")[2];
    upperLinks = within(upperList).getAllByRole("listitem");

    // Login menu opens, revealing two additional list items.
    expect(upperLinks.length).toEqual(8);
    expect(upperLinks[0]).toHaveTextContent(/close/i);
    expect(upperLinks[1]).toHaveTextContent(/go to the catalog/i);
  });

  it("renders the horizontal rule", () => {
    expect(screen.getByRole("separator")).toBeInTheDocument();
  });

  it("renders the UI snapshot correctly", () => {
    const header = renderer.create(<Header isProduction={false} />).toJSON();

    expect(header).toMatchSnapshot();
  });
});

// Skipping because this functionality is temporarily removed.
describe.skip("Patron API call succeeds", () => {
  beforeAll(async () => {
    (Cookies as any).setMockImplementation(() => ({
      get: () => mockLoginCookie,
    }));
    // When the Header mounts, it immediately checks for the cookie value.
    // That is why this mock is here.
    // Cookies.get = jest.fn().mockReturnValue(mockLoginCookie);
    // This is assuming that the sitewide alerts fetch call runs
    // before the patron information call. If the alerts call runs
    // after the patron information call, the wrong response will be
    // returned for each and this test will fail.

    // TODO: Fix the indeterministic fetch calls.
    (global as any).fetch = jest
      .fn()
      // Mock the fetch API call for sitewide alerts.
      .mockReturnValueOnce(
        Promise.resolve({
          status: 200,
          json: () => Promise.resolve(refineryResponse),
        })
      )
      // Mock the fetch API call for patron data (successful).
      .mockReturnValueOnce(
        Promise.resolve({
          status: 200,
          json: () => Promise.resolve(mockResponseData),
        })
      );

    await waitFor(() => render(<Header isProduction={false} />));
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it("displays the logged in view when call succeeds", async () => {
    document.getElementById("chakra-toast-portal")?.remove();

    expect(Cookies.get).toHaveBeenCalledWith("nyplIdentityPatron");

    // The third list is the upper navigation links.
    const upperList = await screen.getAllByRole("list")[2];
    const upperLinks = within(upperList).getAllByRole("listitem");
    const firstListElement = within(upperLinks[0]).getByRole("button");
    let greeting = screen.queryByText(/jane/i);
    let logoutButton = screen.queryByRole("link", { name: /log out/i });

    expect(greeting).not.toBeInTheDocument();
    expect(logoutButton).not.toBeInTheDocument();

    expect(firstListElement).toHaveTextContent(/account/i);

    userEvent.click(firstListElement);

    greeting = screen.queryByText(/jane/i);
    logoutButton = screen.queryByRole("link", { name: /log out/i });

    expect(greeting).toBeInTheDocument();
    expect(logoutButton).toBeInTheDocument();
  });
});

// Skipping because this functionality is temporarily removed.
describe.skip("Patron API call fails", () => {
  beforeAll(async () => {
    (Cookies as any).setMockImplementation(() => ({
      get: () => mockLoginCookie,
    }));
    (global as any).fetch = jest
      .fn()
      // Mock the fetch API call for sitewide alerts.
      .mockReturnValueOnce(
        Promise.resolve({
          status: 200,
          json: () => Promise.resolve(refineryResponse),
        })
      )
      // Mock the fetch API call for patron data (unsuccessful).
      .mockReturnValueOnce(
        Promise.resolve({
          status: 500,
        })
      );

    await waitFor(() => render(<Header isProduction={false} />));
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it("logs warning and displays the logged out view if call fails", async () => {
    const warn = jest.spyOn(console, "warn");
    expect(warn).toHaveBeenCalledWith(
      "NYPL Reservoir Header: There was an error fetching NYPL patron data."
    );

    document.getElementById("chakra-toast-portal")?.remove();

    // If the call fails, the Header should render the logged out state.
    const upperList = await screen.getAllByRole("list")[2];
    const upperLinks = within(upperList).getAllByRole("listitem");
    const firstListElement = within(upperLinks[0]).getByRole("button");

    expect(firstListElement).toHaveTextContent(/log in/i);
  });
});

// Skipping because this functionality is temporarily removed.
describe.skip("Patron API returns wrong data", () => {
  beforeAll(async () => {
    (Cookies as any).setMockImplementation(() => ({
      get: () => mockLoginCookie,
    }));

    (global as any).fetch = jest
      .fn()
      // Mock the fetch API call for sitewide alerts.
      .mockReturnValueOnce(
        Promise.resolve({
          status: 200,
          json: () => Promise.resolve(refineryResponse),
        })
      )
      // Mock the fetch API call for patron data (wrong data).
      .mockReturnValueOnce(
        Promise.resolve({
          status: 200,
          json: () => Promise.resolve({}),
        })
      );

    await waitFor(() => render(<Header isProduction={false} />));
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it("displays the logged out view if wrong data is received", async () => {
    document.getElementById("chakra-toast-portal")?.remove();

    // Without the correct data, the Header should render the logged out state.
    const upperList = await screen.getAllByRole("list")[2];
    const upperLinks = within(upperList).getAllByRole("listitem");
    const firstListElement = within(upperLinks[0]).getByRole("button");

    expect(firstListElement).toHaveTextContent(/log in/i);
  });
});

// Skipping because this functionality is temporarily removed.
describe.skip("Patron API returns expired data, but refreshes the token successfully", () => {
  beforeAll(async () => {
    (Cookies as any).setMockImplementation(() => ({
      get: () => mockLoginCookie,
    }));

    (global as any).fetch = jest
      .fn()
      // Mock the fetch API call for sitewide alerts.
      .mockReturnValueOnce(
        Promise.resolve({
          status: 200,
          json: () => Promise.resolve(refineryResponse),
        })
      )
      // Mock the fetch API call for patron data (token expired).
      .mockReturnValueOnce(
        Promise.resolve({
          status: 401,
          json: () => Promise.resolve(mockExpiredResponseData),
        })
      )
      // Mock the fetch API call to refresh the access token (successful).
      .mockReturnValueOnce(
        Promise.resolve({
          status: 200,
        })
      )
      // Mock the fetch API call for patron data (successful).
      .mockReturnValueOnce(
        Promise.resolve({
          status: 200,
          json: () => Promise.resolve(mockResponseData),
        })
      );

    await waitFor(() => render(<Header isProduction={false} />));
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it("displays logged in view if refresh is successful", async () => {
    expect((global as any).fetch).toHaveBeenNthCalledWith(1, alertsApiUrl);
    expect((global as any).fetch).toHaveBeenNthCalledWith(
      2,
      patronApiUrlWithToken
    );
    expect((global as any).fetch).toHaveBeenNthCalledWith(
      3,
      // Using the development environment.
      tokenRefreshLink(false),
      {
        credentials: "include",
      }
    );
    expect((global as any).fetch).toHaveBeenNthCalledWith(
      4,
      patronApiUrlWithToken
    );

    document.getElementById("chakra-toast-portal")?.remove();

    // The third list is the upper navigation links.
    const upperList = await screen.getAllByRole("list")[2];
    const upperLinks = within(upperList).getAllByRole("listitem");
    const firstListElement = within(upperLinks[0]).getByRole("button");
    let greeting = screen.queryByText(/jane/i);
    let logoutButton = screen.queryByRole("link", { name: /log out/i });

    expect(greeting).not.toBeInTheDocument();
    expect(logoutButton).not.toBeInTheDocument();

    expect(firstListElement).toHaveTextContent(/account/i);

    userEvent.click(firstListElement);

    greeting = screen.queryByText(/jane/i);
    logoutButton = screen.queryByRole("link", { name: /log out/i });

    expect(greeting).toBeInTheDocument();
    expect(logoutButton).toBeInTheDocument();
  });
});

// Skipping because this functionality is temporarily removed.
describe.skip("Patron API returns expired data and cannot refresh the token successfully", () => {
  beforeAll(async () => {
    (Cookies as any).setMockImplementation(() => ({
      get: () => mockLoginCookie,
    }));

    (global as any).fetch = jest
      .fn()
      // Mock the fetch API call for sitewide alerts.
      .mockReturnValueOnce(
        Promise.resolve({
          status: 200,
          json: () => Promise.resolve(refineryResponse),
        })
      )
      // Mock the fetch API call for patron data (token expired).
      .mockReturnValueOnce(
        Promise.resolve({
          status: 401,
          json: () => Promise.resolve(mockExpiredResponseData),
        })
      )
      // Mock the fetch API call to refresh accessToken (unsuccessful)
      .mockReturnValueOnce(
        Promise.resolve({
          status: 500,
        })
      );

    await waitFor(() => render(<Header isProduction={false} />));
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it("displays logged out view if refresh is unsuccessful", async () => {
    const warn = jest.spyOn(console, "warn");
    expect(warn).toHaveBeenCalledWith(
      "NYPL Reservoir Header: There was an error refreshing NYPL patron data."
    );

    document.getElementById("chakra-toast-portal")?.remove();

    // If the access token is expired, the Header should render the logged out state.
    const upperList = await screen.getAllByRole("list")[2];
    const upperLinks = within(upperList).getAllByRole("listitem");
    const firstListElement = within(upperLinks[0]).getByRole("button");

    expect(firstListElement).toHaveTextContent(/log in/i);
  });
});
