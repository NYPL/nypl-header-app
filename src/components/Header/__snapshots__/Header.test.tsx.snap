// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders the UI snapshot correctly 1`] = `
<div
  className="css-0"
>
  <nav
    aria-label="Skip Navigation"
    className="css-1xdhyk6"
  >
    <ul
      className="css-1xdhyk6"
    >
      <li>
        <a
          className="css-1xdhyk6"
          href="#mainContent"
          rel={null}
          target={null}
        >
          Skip to Main Content
        </a>
      </li>
      <li>
        <a
          className="css-1xdhyk6"
          href="https://www.nypl.org/accessibility"
          rel={null}
          target={null}
        >
          Click to learn about accessibility at the Library
        </a>
      </li>
    </ul>
  </nav>
  <header>
    <div
      className="chakra-stack css-1p1m4ay"
    >
      <a
        aria-label="The New York Public Library"
        className="css-1xdhyk6"
        href="https://nypl.org"
        rel={null}
        target={null}
      >
        <svg
          aria-hidden={false}
          aria-label="NYPL Header Logo"
          className="chakra-icon css-1vgifmd"
          fill="none"
          focusable={false}
          role="img"
          viewBox="0 0 136 135"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            NYPL Header Logo
          </title>
          <path
            d="M80.251 43.353c-.87-1.65-3.396-2.04-6.052-1.91a16.55 16.55 0 0 0-4.18.651c-1.48.434-3.484 1.433-2.613 3.473.566 1.39 2.57 1.954 4.006 2.432 1.655.564 3.484.998 4.746 1.736.74.434 1.524 1.563 2.57 1.433 1.044-.13 1.698-2.214 1.915-3.256.305-1.607.175-3.56-.392-4.559Z"
            fill="#000"
          />
          <path
            clipRule="evenodd"
            d="M68.103.112C31.047.112.87 30.155.87 67.492c0 37.336 30.394 67.423 67.275 67.423 37.448 0 67.798-30.043 67.798-67.38C135.901 30.155 105.551.112 68.103.112ZM28.652 114.77a10.757 10.757 0 0 1-2.482-3.039c-5.748-9.16-8.97-41.548-1.568-52.531 2.352-3.864 4.834-2.084 4.006.911-1.437 6.686-.566 10.42 2.7 17.67 3.631 7.609 12.59 17.32 19.008 24.279 2.664 2.887 4.89 5.301 6.117 6.893 7.185 9.29 9.58 16.063 9.536 17.887-.043 1.389-.305 2.3-2.787 2.083-12.932-.998-25.517-6.382-34.53-14.153Zm99.052-30.575.054-.162C138.6 47.52 111.647 4.497 64.837 6.19c-2.961.174-5.4 1.086-3.222 2.779 6.662 5.123 10.232 11.765 11.495 17.8.087.52.349.825 1.045 1.042 3.048.825 10.756 3.864 13.15 5.687.95.723 1.892 1.69 2.868 2.693 1.487 1.528 3.052 3.136 4.84 4.08 3.412 1.8 6.769 3.27 10.024 4.698h.003c1.77.777 3.511 1.54 5.213 2.335 2.134.955 6.053 3.56 5.792 6.642-.041.485-.157.952-.269 1.404-.13.518-.254 1.018-.254 1.505 0 .562.232 1.19.45 1.783.136.368.267.723.334 1.039.446 2.111.563 4.279.348 6.425-.305 2.301-1.045 5.948-2.351 8.075-.134.207-.282.4-.429.592-.375.49-.74.967-.834 1.623a11.233 11.233 0 0 1-.697 2.518c-.13.217-.13.608.131.65 5.138 1.911 10.494 5.34 12.889 7.12 1.22.881 1.83-.951 2.341-2.485Zm-18.975-20.828a13.058 13.058 0 0 1-.13-1.215c0-.359.593-.806 1.046-1.147.095-.072.184-.139.26-.2a13.999 13.999 0 0 1 1.785-1.258c.14-.07.223.055.273.13.013.02.023.035.032.044 1.742 2.604 1.089 9.073-.174 12.156-.131.303-.392.434-.523.217-.653-1.39-1.567-2.649-2.656-3.3a.689.689 0 0 1-.218-.52c.435-1.599.538-3.268.305-4.907ZM68.495 32.24c-7.708-1.867-14.152.564-16.373 2.344-4.18 3.3-7.446 8.64-6.14 16.238.523 3.256 1.524 5.99 3.136 8.248 4.45 6.297 11.622 9.674 18.79 13.048 3.21 1.512 6.42 3.023 9.382 4.796a.632.632 0 0 0 .915-.13l.016-.024c1.604-2.334 2.91-4.235 5.601-5.36 2.43-1 6.123-.808 9.537-.63h.004c.44.023.874.046 1.302.066 3.875.173 8.055-.174 9.449-2.171 1.698-2.388 1.611-7.163-.958-8.64-.742-.415-1.668-.622-2.595-.83-.816-.181-1.633-.364-2.326-.69-1.088-.477-1.916-1.432-2.525-2.734-.48-1.086-.958-2.866-.305-4.038.556-1.066 2.037-.75 3.246-.49l.063.013c2.395.564 6.445 1.128 8.752.13.654-.347-1.785-1.997-2.699-2.431-2.15-1.062-4.607-1.873-7.06-2.684-2.765-.913-5.525-1.825-7.832-3.09-1.545-.858-2.923-1.961-4.307-3.07-1.7-1.36-3.408-2.727-5.447-3.66a74.478 74.478 0 0 0-11.626-4.211ZM37.186 17.391c-1.872-.217-1.915-1.606-1.132-2.127a47.598 47.598 0 0 1 8.753-4.515 4.145 4.145 0 0 1 2.438-.391c2.482.26 6.575 1.91 11.278 5.253 5.617 3.994 10.668 11.635 7.315 11.027a34.483 34.483 0 0 0-3.7-.52c-1.441-.124-2.612-.093-3.57-.067-.723.02-1.325.036-1.83-.02a2.276 2.276 0 0 1-1.263-.695c-4.746-4.255-10.233-6.946-18.288-7.945Zm51.818 59.652a3.828 3.828 0 0 0-1.655 2.214c-.027.09-.069.215-.118.361-.187.558-.473 1.413-.404 1.723.067.37.58.897.969 1.297.113.116.215.221.293.31a35.34 35.34 0 0 0 2.744 2.908 7.351 7.351 0 0 0 1.61 1.26c.604.3 1.44.37 2.348.445.273.022.55.045.832.075l.034.004c3.344.346 7.026.728 8.892-1.176.218-.217.305-.868.348-1.302.206-1.336.871-2.37 1.537-3.406.741-1.154 1.484-2.31 1.599-3.888a6.11 6.11 0 0 0-.349-2.605 2.827 2.827 0 0 0-1.437-1.737c-1.213-.509-2.098.102-2.896.652-.291.2-.57.393-.849.52-1.622.769-4 .934-6.289 1.094l-.503.035c-.296.02-.594.04-.89.058h-.001c-2.234.14-4.432.277-5.815 1.158ZM22.948 26.03c.385-.43.883-.746 1.437-.912a37.386 37.386 0 0 1 23.557 1.997c1.393.608 2.046 1.476 1.132 2.084-1.176.695-4.703 2.996-6.662 4.819-.566.478-1.568.521-2.177.434l-.246-.032c-7.157-.935-11.025-1.44-21.875 2.68-1.698.695-3.222-.26-2.438-1.563 1.437-2.648 4.833-7.033 7.272-9.507ZM8.36 53.816a2.946 2.946 0 0 0-.653 1.606c-2.787 13.502-.566 24.79 2.09 31.997.61 1.563 3.048.955 2.961-2.388.017-.605.023-1.275.03-2.002.082-9.085.246-27.125 21.09-39.546 1.262-.738 2.35-1.476 1.828-2.778-.305-.738-1.655-.868-3.484-.608-14.064 2.475-23.121 12.59-23.862 13.72Zm27.738 17.28c-3.963-9.639-1.393-18.279.392-21.578.87-1.65 1.741-2.735 2.874-2.431 1.219.347 1.524 1.606 1.698 3.907 1.263 19.233 12.932 25.702 27.563 32.518 11.583 5.904 24.733 12.633 35.75 28.957 1.654 2.345 1.916 3.777.043 5.123-6.705 4.559-12.802 7.771-21.728 9.855-1.35.304-2.743.087-3.919-2.995-6.742-17.265-15.964-24.728-24.497-31.632-7.174-5.805-13.86-11.216-18.176-21.725Zm73.067 38.334c.609 1.172 1.698 2.649 3.57.782 3.223-3.17 7.795-9.291 9.319-12.243.035-.077.074-.154.115-.235.258-.511.56-1.11.146-1.936-.74-1.52-4.877-6.165-7.664-8.292l-.033-.025c-1.086-.823-2.55-1.933-3.929-2.32a3.246 3.246 0 0 0-1.089-.13c-1.025.118-1.158.592-1.561 2.034l-.137.484c-.74 2.518-2.177 4.298-3.44 4.82a6.957 6.957 0 0 1-3.745.607c-1.088-.087-1.654.912-.914 1.65 4.136 4.515 6.705 9.204 9.362 14.804Z"
            fill="#000"
            fillRule="evenodd"
          />
          <path
            d="M71.456 47.998c1.654.565 3.483.999 4.746 1.737.74.434 1.524 1.563 2.569 1.433 1.045-.13 1.698-2.214 1.916-3.256.305-1.563.13-3.517-.392-4.515-.87-1.65-3.397-2.041-6.053-1.91-1.415.037-2.82.256-4.18.65-1.48.435-3.483 1.433-2.612 3.474.566 1.302 2.569 1.91 4.005 2.388Z"
            fill="#000"
          />
        </svg>
      </a>
      <div
        className="css-17xejub"
      />
      <div
        className="css-k008qs"
      >
        <div
          className="css-0"
        >
          <div
            data-focus-guard={true}
            style={
              {
                "height": "0px",
                "left": "1px",
                "overflow": "hidden",
                "padding": 0,
                "position": "fixed",
                "top": "1px",
                "width": "1px",
              }
            }
            tabIndex={-1}
          />
          <div
            data-focus-guard={true}
            style={
              {
                "height": "0px",
                "left": "1px",
                "overflow": "hidden",
                "padding": 0,
                "position": "fixed",
                "top": "1px",
                "width": "1px",
              }
            }
            tabIndex={-1}
          />
          <div
            data-focus-lock-disabled="disabled"
            onBlur={[Function]}
            onFocus={[Function]}
          >
            <button
              aria-label="My Account"
              className="chakra-button css-1xdhyk6"
              id="loginButton"
              onClick={[Function]}
              type="button"
            >
              <svg
                aria-hidden={true}
                className="chakra-icon css-1vgifmd"
                focusable={false}
                role="img"
                viewBox="0 0 32 33"
                xmlns="http://www.w3.org/2000/svg"
              >
                <title>
                  Log in to your account
                </title>
                <path
                  d="M16.106 13.732c3.194 0 5.784-3.074 5.784-6.866S19.3 0 16.106 0s-5.784 3.074-5.784 6.866 2.59 6.866 5.784 6.866ZM29.087 32.003 2.74 31.992a2.9 2.9 0 0 1-2.684-2.335 2.909 2.909 0 0 1 .03-1.266l3.478-9.2a2.999 2.999 0 0 1 1.903-1.675c3.271-.563 6.58-.874 9.899-.93 4.166 0 8.292.64 10.018.764a3.11 3.11 0 0 1 2.788 1.57l3.706 9.404a2.86 2.86 0 0 1-1.49 3.395 2.849 2.849 0 0 1-1.3.284Z"
                />
              </svg>
            </button>
          </div>
          <div
            data-focus-guard={true}
            style={
              {
                "height": "0px",
                "left": "1px",
                "overflow": "hidden",
                "padding": 0,
                "position": "fixed",
                "top": "1px",
                "width": "1px",
              }
            }
            tabIndex={-1}
          />
        </div>
        <a
          aria-label="NYPL Locations Near Me"
          className="css-1xdhyk6"
          href="https://nypl.org/locations"
          rel={null}
          target={null}
        >
          <svg
            aria-hidden={true}
            className="chakra-icon css-1vgifmd"
            focusable={false}
            role="img"
            viewBox="0 0 32 32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              NYPL Locator
            </title>
            <path
              d="M16 11.594a2.004 2.004 0 1 1-2.004 2.004A2.004 2.004 0 0 1 16 11.594m0-1.687a3.691 3.691 0 1 0 3.69 3.69A3.695 3.695 0 0 0 16 9.908"
            />
            <path
              d="M27.237 10.695a11.537 11.537 0 0 0-2.93-5.215c-.032-.032-.058-.068-.09-.1-.033-.033-.07-.06-.102-.093a11.529 11.529 0 0 0-5.221-2.935 11.332 11.332 0 0 0-5.789 0 11.526 11.526 0 0 0-5.22 2.935c-.034.033-.07.06-.103.093-.032.032-.058.068-.09.1a11.528 11.528 0 0 0-2.93 5.215 11.693 11.693 0 0 0 0 5.806 11.527 11.527 0 0 0 2.93 5.215c.032.033.058.068.09.099.033.034.07.06.103.094L16 30.024l8.115-8.115c.033-.033.07-.06.102-.094.032-.031.058-.066.09-.099a11.536 11.536 0 0 0 2.93-5.215 11.693 11.693 0 0 0 0-5.806m-4.444 10.25-3.899 3.899L16 27.738l-2.895-2.894-3.899-3.9-.552-.552a10.002 10.002 0 0 1-.001-13.588l.554-.553a9.971 9.971 0 0 1 13.585 0l.555.554a10.002 10.002 0 0 1 0 13.587Z"
            />
          </svg>
        </a>
        <div
          className="css-0"
        >
          <div
            data-focus-guard={true}
            style={
              {
                "height": "0px",
                "left": "1px",
                "overflow": "hidden",
                "padding": 0,
                "position": "fixed",
                "top": "1px",
                "width": "1px",
              }
            }
            tabIndex={-1}
          />
          <div
            data-focus-guard={true}
            style={
              {
                "height": "0px",
                "left": "1px",
                "overflow": "hidden",
                "padding": 0,
                "position": "fixed",
                "top": "1px",
                "width": "1px",
              }
            }
            tabIndex={-1}
          />
          <div
            data-focus-lock-disabled="disabled"
            onBlur={[Function]}
            onFocus={[Function]}
          >
            <button
              aria-expanded={null}
              aria-haspopup="true"
              aria-label="Open Search"
              className="chakra-button css-1xdhyk6"
              id="searchButton"
              onClick={[Function]}
              type="button"
            >
              <span>
                <svg
                  aria-hidden={true}
                  className="chakra-icon css-1vgifmd"
                  focusable={false}
                  role="img"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <title>
                    Open Search
                  </title>
                  <path
                    clipRule="evenodd"
                    d="M18 10a8 8 0 1 1-16 0 8 8 0 0 1 16 0zm-1.671 7.743A9.958 9.958 0 0 1 10 20C4.477 20 0 15.523 0 10S4.477 0 10 0s10 4.477 10 10a9.958 9.958 0 0 1-2.257 6.329l5.96 5.96a1 1 0 0 1-1.414 1.414l-5.96-5.96z"
                    fillRule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            data-focus-guard={true}
            style={
              {
                "height": "0px",
                "left": "1px",
                "overflow": "hidden",
                "padding": 0,
                "position": "fixed",
                "top": "1px",
                "width": "1px",
              }
            }
            tabIndex={-1}
          />
        </div>
        <div
          className="css-0"
        >
          <div
            data-focus-guard={true}
            style={
              {
                "height": "0px",
                "left": "1px",
                "overflow": "hidden",
                "padding": 0,
                "position": "fixed",
                "top": "1px",
                "width": "1px",
              }
            }
            tabIndex={-1}
          />
          <div
            data-focus-guard={true}
            style={
              {
                "height": "0px",
                "left": "1px",
                "overflow": "hidden",
                "padding": 0,
                "position": "fixed",
                "top": "1px",
                "width": "1px",
              }
            }
            tabIndex={-1}
          />
          <div
            data-focus-lock-disabled="disabled"
            onBlur={[Function]}
            onFocus={[Function]}
          >
            <button
              aria-expanded={null}
              aria-haspopup="true"
              aria-label="Open Navigation"
              className="chakra-button css-1xdhyk6"
              id="mobileNav-btn"
              onClick={[Function]}
              type="button"
            >
              <svg
                aria-hidden={true}
                className="chakra-icon css-1vgifmd"
                focusable={false}
                role="img"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <title>
                  utilityHamburger icon
                </title>
                <path
                  clipRule="evenodd"
                  d="M3 17a1 1 0 0 1 1-1h16a1 1 0 1 1 0 2H4a1 1 0 0 1-1-1ZM3 12a1 1 0 0 1 1-1h16a1 1 0 1 1 0 2H4a1 1 0 0 1-1-1ZM3 7a1 1 0 0 1 1-1h16a1 1 0 1 1 0 2H4a1 1 0 0 1-1-1Z"
                  fillRule="evenodd"
                />
              </svg>
            </button>
          </div>
          <div
            data-focus-guard={true}
            style={
              {
                "height": "0px",
                "left": "1px",
                "overflow": "hidden",
                "padding": 0,
                "position": "fixed",
                "top": "1px",
                "width": "1px",
              }
            }
            tabIndex={-1}
          />
        </div>
      </div>
    </div>
    <hr
      className="css-mvtkpx"
    />
  </header>
</div>
`;
